File: 67-add-binary/add-binary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function(a, b) {
    let l1 = a.length
    let l2 = b.length
    if(l1 < l2){
        let temp = l2
        l2 = l1
        l1 = temp
        let tempS = b
        b = a
        a = tempS
    }
    // console.log(l1, l2)
    l2 -= 1
    let carry = false;
    let ans = "";
    for(let i = l1-1; i >= 0; i--){
        if(l2 < 0){
          if(a[i] == '0'){
            if(carry){
                ans += '1'
                carry = false
            }else{
                ans += '0'
            }
          }else{
            if(carry){
                ans += '0';
            }else{
                ans += '1';
            }
          }
        }else{
            if(a[i] == '1' && b[l2] == '1'){
                if(carry){
                    ans += '1';
                }else{
                    ans += '0';
                    carry = true;
                }
            }else if(a[i] == '0' && b[l2] == '0'){
                if(carry){
                    ans += '1';
                    carry = false;
                }else{
                    ans += '0';
                }
            }else{
                if(carry){
                    ans += '0';
                }else{
                    ans += '1';
                }
            }
        }
        l2--
        // console.log(ans, carry)
    }
    if(carry){
        ans += '1'
    }
    ans = ans.split('').reverse().join('')
    return ans;
};

File: 67-add-binary/add-binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        //its not needed but let's do it stupidly
        string ans = "";
        reverse(a.begin(), a.end());
        reverse(b.begin(), b.end());
        int l = max(a.length(), b.length());
        int i = 0;
        if(a.length() < b.length()){
            swap(a,b);
        }
        int carry = 0;
        while(i < l){
            cout<<"i -> "<<i<<endl;
            cout<<ans<<" ";
            int no = a[i] -  '0';
            if(i > b.length()-1){
                if(no){
                    if(carry){
                        ans.push_back('0');
                    }else{
                        ans.push_back('1');
                    }
                }else{
                    if(carry){
                        ans.push_back('1');
                        carry = 0;
                    }else{
                        ans.push_back('0');
                    }
                }
            }else{
                int no1 = a[i] - '0';
                int no2 = b[i] - '0';
                cout<<" -> "<<no1<<" "<<no2<<endl;
                int total = no1 + no2 + carry;
                if(total == 0){
                    ans.push_back('0');
                }else if(total == 1){
                    ans.push_back('1');
                    carry = 0;
                }else if(total == 2){
                    ans.push_back('0');
                    carry = 1;
                }else if(total == 3){
                    ans.push_back('1');
                    carry = 1;
                }
            }
            i++;
            cout<<ans<<endl;
        }
            if(carry){
                ans.push_back('1');
            }
            if(ans.length() > 1 and ans.back() == '0'){
                ans.pop_back();
            }

            reverse(ans.begin(), ans.end());
            return ans;
    }
};